File: algo-c-to-_/src/gcd.py
#
#	from src/gcd.c
#
#	int gcd(int, int) ; recursive	to	gcdR
#	int gcd(int, int) ; loop	to	gcdL
#	int ngcd(int, int[])		to	ngcdL
#	ngcdL				to	ngcdR
#

def gcdR(x, y):
    if y == 0:
        return x
    else:
        return gcdR(y, x % y)

def gcdL(x, y):
    while y != 0:
        x, y = y, x % y
    return x

def ngcdL(a):
    if not type(a) in (type(()), type([])) or len(a) == 0:
        return

    d = a[0]
    for v in a[1:]:
        if d == 1:
            break
        d = gcdL(v, d)
    return d

def ngcdR(a):
    if not type(a) in (type(()), type([])) or len(a) == 0:
        return

    def iter(i, d):
        if i >= len(a) or d == 1:
            return d
        else:
            return iter(i+1, gcdR(a[i], d))

    return iter(1, a[0])

File: porting-from-algo-c/src/gcd.lua
--
--	from src/gcd.c
--
--	int gcd(int, int) ; recursive	to	gcdR
--	int gcd(int, int) ; loop	to	gcdL
--	int ngcd(int, int[])		to	ngcdL
--	ngcdL				to	ngcdR
--

local isTbl = require '_helper'.isTbl

local function gcdR(x, y)
	if y == 0 then
		return x
	else
		return gcdR(y, x % y)
	end
end

local function gcdL(x0, y0)
	local x, y = x0, y0
	while y ~= 0 do
		x, y = y, x % y
	end
	return x
end

local function ngcdL(a)
	assert(isTbl(a), "ngcdL needs a table (of numbers).")
	local d = a[1]
	for i=2,#a do
		if d == 1 then break end
		d = gcdL(a[i], d)
	end
	return d
end

local function ngcdR(a)
	assert(isTbl(a), "ngcdR needs a table (of numbers).")
	function rec(i, d)
		if i > #a or d == 1 then
			return d
		else
			return rec(i+1, gcdR(a[i], d))
		end
	end
	return rec(1, a[1])
end

return {
	gcdR = gcdR,
	gcdL = gcdL,
	ngcdL = ngcdL,
	ngcdR = ngcdR
}
